base_layer/core/src/transactions/coinbase_builder.rs

```diff
@@ -173,113 +162,167 @@ impl CoinbaseBuilder {
-        let spending_key_hash_bytes = spending_key_hash.as_ref();
-
-        let sender_offset_private_key =
-            PrivateKey::from_bytes(spending_key_hash_bytes).map_err(|_| CoinbaseBuildError::InvalidSenderOffsetKey)?;
-        let sender_offset_public_key = PublicKey::from_secret_key(&sender_offset_private_key);
-        let covenant = self.covenant;
-
+        let public_nonce_id = self
+            .key_manager
+            .get_next_key_id(CoreKeyManagerBranch::Nonce.get_branch_key())
+            .await?;
+
+        let public_nonce = self.key_manager.get_public_key_at_key_id(&public_nonce_id).await?;
+        let public_spend_key = self.key_manager.get_public_key_at_key_id(&spending_key_id).await?;
+        let public_script_key = self.key_manager.get_public_key_at_key_id(&script_key_id).await?;
+
+        let kernel_signature = self
+            .key_manager
+            .get_partial_kernel_signature(
+                &spending_key_id,
+                &public_nonce_id,
+                &public_nonce,
+                &public_spend_key,
+                &kernel_version,
+                &kernel_message,
+                &metadata.kernel_features,
+                TxoType::Output,
+            )
+            .await?;
+        let excess_public_key = self.key_manager.get_public_key_at_key_id(&spending_key_id).await?;
+        let excess = Commitment::from_public_key(&excess_public_key);
+        // generate tx details
+        let value: u64 = total_reward.into();
+        let output_features = OutputFeatures::create_coinbase(height + constants.coinbase_lock_height(), self.extra);
         let encrypted_data = self
-            .recovery_data
-            .as_ref()
-            .map(|rd| EncryptedData::encrypt_data(&rd.encryption_key, &commitment, total_reward, &spending_key))
-            .transpose()
-            .map_err(|_| CoinbaseBuildError::ValueEncryptionFailed)?
-            .unwrap_or_default();
-
+            .key_manager
+            .encrypt_data_for_recovery(&spending_key_id, &None, total_reward.into())
+            .await?;
         let minimum_value_promise = MicroTari::zero();
 
-        let metadata_sig = TransactionOutput::create_metadata_signature(
-            TransactionOutputVersion::get_current_version(),
-            total_reward,
-            &spending_key,
+        let output_version = TransactionOutputVersion::get_current_version();
+        let metadata_message = TransactionOutput::build_metadata_signature_message(
+            &output_version,
             &script,
             &output_features,
-            &sender_offset_private_key,
             &covenant,
             &encrypted_data,
             minimum_value_promise,
-        )
-        .map_err(|e| CoinbaseBuildError::BuildError(e.to_string()))?;
+        );
+
+        // because the coinbase does not have a spending input with a script key, we use the output spending key as
```

Comment:
@AaronFeickert is this secure?

Reply:
it will be good to know as thats how its currently done:
https://github.com/tari-project/tari/blob/f3833c9fc46d77fddaa7a23ef1d53ba9d860182a/base_layer/wallet/src/output_manager_service/service.rs#LL1024C10-L1024C10